### PR TITLE
registry_proxy_model: count repository size when caching images (PROJQUAY-3464)

### PR DESCRIPTION
**Pull request by flavianmissi**
also check if the org is within its quota limits before caching an image
from the upstream registry.

### CODE REVIEW

File: data/registry_model/registry_proxy_model.py [hunk 1]
```diff
@@ -469,6 +483,7 @@ def _download_blob(self, repo_ref: RepositoryReference, digest: str) -> int:
         Download blob from upstream registry and perform a monolitic upload to
         Quay's own storage.
         """
+        self._enforce_repository_quota(repo_ref)
```
Comment (kwestpharedhat) [inline data/registry_model/registry_proxy_model.py:486]: running this prior to actually doing the download will result in stale data

File: data/registry_model/test/test_registry_proxy_model.py [hunk 2]
```diff
@@ -664,7 +664,7 @@ def test_renew_manifest_and_parent_tag_when_manifest_is_child_of_manifest_list(
         assert updated_tag.lifetime_end_ms > manifest_tag.lifetime_end_ms
         assert updated_list_tag.lifetime_end_ms > manifest_list_tag.lifetime_end_ms
 
-    def test_update_manifest_bytes_when_manifest_is_placeholder(
+    def test_update_manifest_bytes_and_layers_compressed_size_when_manifest_is_placeholder(
```
Comment (kwestpharedhat) [inline data/registry_model/test/test_registry_proxy_model.py:667]: is there a way to shorten this function name? This name by itself is over 80 characters which violates most style guides.

File: endpoints/v2/__init__.py [hunk 3]
```diff
@@ -61,8 +63,19 @@ def handle_readonly(ex):
 
 
 @v2_bp.app_errorhandler(UpstreamRegistryError)
-def handle_proxy_cache_errors(error):
+def handle_proxy_cache_error(error):
```
Comment (kwestpharedhat) [inline endpoints/v2/__init__.py:66]: This and the following function are very similar perhaps make the response formatting a function and reuse that way.

File: data/registry_model/registry_proxy_model.py [hunk 4]
```diff
@@ -266,6 +277,7 @@ def _create_and_tag_manifest(
         or the retrieved manifest is invalid (for docker manifest schema v1).
         """
         self._proxy.manifest_exists(manifest_ref, ACCEPTED_MEDIA_TYPES)
+        self._enforce_repository_quota(repo_ref)
```
Comment (kwestpharedhat) [inline data/registry_model/registry_proxy_model.py:280]: any form of force caching should be the final thing to do. If you need a check to ensure the quota isnt exceeded at the beginning do not use force cache for performance and server load reduction.

**Review by kwestpharedhat — CHANGES_REQUESTED** [review]
Changes requested

File: data/registry_model/test/test_registry_proxy_model.py (hunk 2 above)
Comment (flavianmissi) [reply data/registry_model/test/test_registry_proxy_model.py:667]: done

File: endpoints/v2/__init__.py (hunk 3 above)
Comment (flavianmissi) [reply endpoints/v2/__init__.py:66]: good point - done

File: data/registry_model/registry_proxy_model.py (hunk 4 above)
Comment (kwestpharedhat) [reply data/registry_model/registry_proxy_model.py:280]: that order of operations would be correct yes. 

Also yes you may be correct that the order of operations in manifest v2 may be incorrect I will re-evaluate that flow.

**Review by kwestpharedhat — APPROVED** [review]

**Pull request merged**